addons/website_sale_loyalty/controllers/main.py

```diff
@@ -130,3 +130,21 @@ def cart_update_json(self, *args, set_qty=None, **kwargs):
             # and does not follow the request's context
             request.website = request.website.with_context(website_sale_loyalty_delete=True)
         return super().cart_update_json(*args, set_qty=set_qty, **kwargs)
+
+
+class PaymentPortal(main.PaymentPortal):
+
+    def _validate_transaction_for_order(self, transaction, sale_order_id):
+        """Update programs & rewards before finalizing transaction.
+
+        :param payment.transaction transaction: The payment transaction
+        :param int order_id: The sales order to pay, as a `sale.order` id
+        :raise: ValidationError the order amount changed after updating rewards
+        """
+        super()._validate_transaction_for_order(transaction, sale_order_id)
+        order = request.website.sale_get_order()
+        if order and order.id == sale_order_id:
+            initial_amount = order.amount_total
+            order._update_programs_and_rewards()
+            if initial_amount != order.amount_total:
+                raise ValidationError(_("Cannot process payment: applied promotion has expired."))
```

Comment:
We should maybe be a bit more generic, like "blablabla promotion have changed, might be due to one being expired"?
Because it might not be due to one that expired... Maybe the promotion was deleted or whatever? :thinking: